test/e2e/expirable-airdrop.spec.ts

```diff
@@ -0,0 +1,188 @@
+    const ExpirableAirdrop = await ethers.getContractFactory('ExpirableAirdrop');
+    expirableAirdrop = await ExpirableAirdrop.deploy(deployer.address, token.address, expirationTimestamp, root);
+  });
+
+  when('deposit', () => {
+    let toDeposit: BigNumber;
+
+    given(async () => {
+      toDeposit = toUnit(airdropAmount * (accounts.length + 1));
+
+      // transfer some tokens
+      await token.mint(deployer.address, toDeposit);
+    });
+
+    then('airdrop is deposited', async () => {
+      // approve
+      await token.approve(expirableAirdrop.address, ethers.constants.MaxUint256);
+
+      // deposit
+      await expirableAirdrop.depositTokens(toDeposit);
+
+      // save snapshot for later
+      snapshot = await takeSnapshot();
+
+      // expect
+      expect(await token.balanceOf(expirableAirdrop.address)).equal(toDeposit);
+    });
+  });
+
+  when('claim', () => {
+    let alice: any;
+    let leaf: any;
+    let proof: any;
+
+    given(async () => {
+      alice = accounts[1];
+      leaf = getLeaf(alice.address, toUnit(airdropAmount));
+      proof = tree.getHexProof(leaf);
+    });
+
+    then('airdrop claimed', async () => {
+      // user claims airdrop
+      await expirableAirdrop.connect(alice).claim(alice.address, toUnit(airdropAmount), proof);
+
+      // expect
+      expect(await token.balanceOf(alice.address)).equal(toUnit(airdropAmount));
+    });
+
+    then('revert if already claimed', async () => {
+      await expect(expirableAirdrop.connect(alice).claim(alice.address, toUnit(airdropAmount), proof)).to.be.revertedWithCustomError(
+        expirableAirdrop,
+        'AlreadyClaimed'
+      );
+    });
+
+    then('revert if no airdrop', async () => {
+      await impersonateAccount(randomAddress);
+      let random: any = await ethers.getSigner(randomAddress);
+      await setBalance(randomAddress, toUnit(10000));
+      let _leaf = getLeaf(randomAddress, toUnit(airdropAmount));
+      let _proof = tree.getHexProof(_leaf);
+
+      await expect(expirableAirdrop.connect(random).claim(randomAddress, toUnit(airdropAmount), _proof)).to.be.revertedWithCustomError(
+        expirableAirdrop,
+        'NotInMerkle'
+      );
+    });
+  });
+
+  when('claim and transfer', () => {
+    let bob: any;
+    let carl: any;
+    let receiver: any;
+    let leaf: any;
+    let proof: any;
+
+    given(async () => {
+      bob = accounts[2];
+      carl = accounts[3];
+      await impersonateAccount(randomAddress2);
+      receiver = await ethers.getSigner(randomAddress2);
+      await setBalance(randomAddress, toUnit(10000));
+      leaf = getLeaf(bob.address, toUnit(airdropAmount));
+      proof = tree.getHexProof(leaf);
+    });
+
+    then('airdrop claimed', async () => {
+      // user claims airdrop
+      await expirableAirdrop.connect(bob).claimAndTransfer(randomAddress2, toUnit(airdropAmount), proof);
+
+      // expect
+      expect(await token.balanceOf(randomAddress2)).equal(toUnit(airdropAmount));
+    });
+
```

Comment:
1. Could we add a test that reverts if already claimed?
2. Could we add a test that reverts if the user doesn't have something to claim?

I know that technically that was tested on `claim`, but I think it takes only a little time, and I prefer if the test doesn't know that they are implemented with the same internal function